Add additional close statuses in ManagedWebSocket

IDisposable opened the pull request:
Add ServiceRestart (1012), TryAgainLater (1013), and BadGateway (1014) to the list of `WebSocketCloseStatus` values and allow them to be used as valid WebSocket close statuses so we don't reject the close and discard the status description by adding them to the private `IsValueCloseStatus` method switch statement declaring them as valid `true`.

Fixes https://github.com/dotnet/runtime/issues/82602

# Description

The current implementation of `ManagedWebSocket` explicitly declares some `WebSocketCloseStatus` values as "acceptable" reasons to close the socket. For those, the information status description is extracted and made available. For all other values, the close status is rejected and the closing information is discarded.

This means that things like `BadGateway` (1014), or `ServiceRestart` (1012),  or `TryAgainLater` (1013)  will be declared invalid by `ManagedWebSocket.IsValidCloseStatus` and thus not handled cleanly even though they could happen after a web socket is completely setup. These codes are documented [here](https://developer.mozilla.org/en-US/docs/Web/API/CloseEvent/code) as valid server-initiated close reasons.

# Customer Impact

Lost information and ragged closing of a web socket when a server-side or proxy closes because of any of the previously rejected values:

*  ServiceRestart = 1012  // indicates that the server / service is restarting.
*  TryAgainLater = 1013   // indicates that a temporary server condition forced blocking the client's request.
*  BadGateway = 1014       // indicates that the server acting as gateway received an invalid response

This codes are documented [here](https://kapeli.com/cheat_sheets/WebSocket_Status_Codes.docset/Contents/Resources/Documents/index) and [here as IANA registered](https://netty.io/4.1/api/io/netty/handler/codec/http/websocketx/WebSocketCloseStatus.html) and in the [Mozilla docs](https://developer.mozilla.org/en-US/docs/Web/API/CloseEvent/code) 

# Regression

No

# Testing

Added test cases for all three to new WebSocketCloseTests.cs file with no regressions in current tests.

# Risk

This **does not** change the public enum of `WebSocketCloseStatus` as we don't want to invoke public review and the values are not mentioned in the RFC.

# Package authoring signed off?

N/A

IDisposable committed:
Allow non RFC  close statuses in ManagedWebSocket

Add ServiceRestart (1012), TryAgainLater (1013), and BadGateway (1014) to the list of `WebSocketCloseStatus` values and allow them to be used as valid WebSocket close statuses so we don't reject the close and discard the status description by adding them to the private `IsValueCloseStatus` method switch statement declaring them as valid `true`. 
These codes are documented [here as IANA registered codes](https://developer.mozilla.org/en-US/docs/Web/API/CloseEvent/code) as valid server-initiated close reasons.

Fixes Issue https://github.com/dotnet/runtime/issues/82602

IDisposable commented:
This replaces https://github.com/dotnet/runtime/pull/83713

IDisposable commented:
@wfurt and @CarnaViire I have redone this without the public enum change. Let me know if there's a better test (I don't know the WebSockets protocol at all, so I didn't build up native packets, sorry).

IDisposable committed:
Cleanup comment format

Co-authored-by: MartyIX <203266+MartyIX@users.noreply.github.com>

wfurt reviewed (approved):
LGTM. thanks @IDisposable 

We _could_ make the extra codes `private static const` inside of `ManagedWebSocket.cs`  but I don't have strong feelings about it. The extra comment seems sufficient to me.

IDisposable commented:
That's an interesting idea @wfurt. Could/would it be an `internal enum` or bundle of `static const int` values in an adjacent .cs file be more appropriate since we need the constants in both `ManagedWebSocket` and in the test? I've not been into the runtime code until very recently so not sure what the running practices are... I could craft it up both ways on a couple branches for comment if desired.

wfurt commented:
I would wait for feedback from others...

CarnaViire commented on src/libraries/System.Net.WebSockets/tests/WebSocketCloseTests.cs:
```diff
@@ -0,0 +1,94 @@
+
+        public CancellationToken CancellationToken => _cancellation?.Token ?? default;
+
+        public static object[][] ClosesData = {
```
```suggestion
        public static object[][] CloseStatuses = {
```

IDisposable replied:
I'll update all three :)

CarnaViire commented on src/libraries/System.Net.WebSockets/tests/WebSocketCloseTests.cs:
```diff
@@ -0,0 +1,94 @@
+
+        [Theory]
+        [MemberData(nameof(ClosesData))]
+        public async Task ReceivesClose(WebSocketCloseStatus closeStatus)
```
```suggestion
        public async Task ReceiveAsync_ValidCloseStatus_Success(WebSocketCloseStatus closeStatus)
```

IDisposable replied:
Nice! Thanks.

CarnaViire commented on src/libraries/System.Net.WebSockets/tests/WebSocketCloseTests.cs:
```diff
@@ -0,0 +1,94 @@
+                Assert.NotNull(server);
+                Assert.NotNull(client);
+
+                string closeStatusDescription = "closeStatus " + closeStatus.ToString();
```
This string is sent below as a *text* message, so using `closeStatusDescription` name here is confusing. While I think sending and receiving *Text* frames is *not* really necessary here, as the goal is to test a *Close* frame, but if you'd like to do so, I'd suggest having a separate variable named `message` or something like this.

IDisposable replied:
I was mirroring the final argument declaration in the `WebSocketReceiveResult` class but happy to change to `message` if preferred!

```csharp
public WebSocketReceiveResult(int count, WebSocketMessageType messageType, bool endOfMessage, WebSocketCloseStatus? closeStatus, string? closeStatusDescription);

```

IDisposable replied:
Did a cleanup pass now that I better understand the testing of this and in the process this was renamed to `serverMessage`

IDisposable committed:
Rename test data to CloseStatuses

Co-authored-by: Natalia Kondratyeva <knatalia@microsoft.com>

IDisposable committed:
Rename test method to follow naming convention.

Co-authored-by: Natalia Kondratyeva <knatalia@microsoft.com>

IDisposable commented on src/libraries/System.Net.WebSockets/tests/WebSocketCloseTests.cs:
```diff
@@ -0,0 +1,94 @@
+
+                var received = await ReceiveAsync(client);
+                Assert.Equal(closed, received);
+            }
```
See, I said I wasn't sure how to really test this... I can try adding another await.

IDisposable replied:
Now that I better understand how to test this, I've rewritten the tests to send the opening message (without `EndMessage`) and the server-side close, then await both the message and then the close.

IDisposable committed:
Addressed PR feedback

Finished rename of CloseStatuses test data
Renamed `closeStatusDescription` to `serverMessage`
Send hello message and close status then await both responses and
check they are as expected. This necessitated switching to the `ReceiveAsync` that accepts an `ArraySegment`.
Explicitly typed `var`s
Inlined helper methods (for clarity)

IDisposable commented:
Thanks for all the feedback @MartyIX, @wfurt, and @CarnaViire.
I think this ready for another round of feedback or approval. 
Thanks for teaching me how to fish!

CarnaViire commented on src/libraries/System.Net.WebSockets/tests/WebSocketCloseTests.cs:
```diff
@@ -0,0 +1,89 @@
+            WebSocketTestStream stream = new();
+            Encoding encoding = Encoding.UTF8;
+
+            using (WebSocket server = WebSocket.CreateFromStream(stream, true, null, TimeSpan.FromSeconds(3)))
+            using (WebSocket client = WebSocket.CreateFromStream(stream.Remote, false, null, TimeSpan.FromSeconds(3)))
```
```suggestion
            using (WebSocket server = WebSocket.CreateFromStream(stream, isServer: true, null, TimeSpan.FromSeconds(3)))
            using (WebSocket client = WebSocket.CreateFromStream(stream.Remote, isServer: false, null, TimeSpan.FromSeconds(3)))
```

IDisposable committed:
Rename local for per PR feedback

Swapping back to a distinct and more appropriately named variable for the `closeStatusDescription`

Co-authored-by: Natalia Kondratyeva <knatalia@microsoft.com>

IDisposable committed:
Label the boolean for isServer flag

Co-authored-by: Natalia Kondratyeva <knatalia@microsoft.com>

CarnaViire commented on src/libraries/System.Net.WebSockets/tests/WebSocketCloseTests.cs:
```diff
@@ -0,0 +1,89 @@
+                Assert.NotNull(client);
+
+                // send something
+                string hello = "Testing " + closeStatus.ToString();
+                byte[] sendBytes = encoding.GetBytes(hello);
+                await server.SendAsync(sendBytes.AsMemory(), WebSocketMessageType.Text, WebSocketMessageFlags.DisableCompression, CancellationToken);
```
```suggestion
                string hello = "Testing " + closeStatus.ToString();
                byte[] sendBytes = encoding.GetBytes(hello);
                await server.SendAsync(sendBytes.AsMemory(), WebSocketMessageType.Text, WebSocketMessageFlags.EndOfMessage, CancellationToken);
```

IDisposable replied:
I thing we need the `DisableCompression` to ensure we don't get compressed data out (I borrowed that from elsewhere).

As for the `EndOfMessage`, I _intentionally_ switch away from that being sent that way so we are testing a path more reflective of the real world, where we're sending and sending and then hang-up. If that's a bad understanding (and it easliy can be, this is NOT my area), I would happily apply this change, but I would rather it still have the compression off like it was originally: 
```csharp
await server.SendAsync(sendBytes.AsMemory(),
                       WebSocketMessageType.Text,
                       WebSocketMessageFlags.DisableCompression | WebSocketMessageFlags.EndOfMessage,
                       CancellationToken);
```

CarnaViire replied:
Compression is opt-in, so unless you created your WebSocket instance with [`DangerousDeflateOptions`](https://learn.microsoft.com/en-us/dotnet/api/system.net.websockets.websocketcreationoptions.dangerousdeflateoptions?view=net-7.0#system-net-websockets-websocketcreationoptions-dangerousdeflateoptions) creation option (in "real world" it would start with both client *and* server opting in compression and negotiating it during initial handshake) -- this flag would be ignored, as the compression is off.

Also, re "we don't get compressed data out" -- even if compression IS enabled, it is completely transparent for end-users, it will get automatically decompressed upon arrival.

As for EndOfMessage, I'm fine either way.

IDisposable replied:
I told you it wasn't my area :) 

I'll update that when I rebase everything... in process now.

CarnaViire commented on src/libraries/System.Net.WebSockets/tests/WebSocketCloseTests.cs:
```diff
@@ -0,0 +1,89 @@
+                await server.SendAsync(sendBytes.AsMemory(), WebSocketMessageType.Text, WebSocketMessageFlags.DisableCompression, CancellationToken);
+
+                // and then server-side close with the test status
+                string serverMessage = "CloseStatus " + closeStatus.ToString();
+                await server.CloseOutputAsync(closeStatus, serverMessage, CancellationToken);
```
```suggestion
                string closeStatusDescription = "CloseStatus " + closeStatus.ToString();
                await server.CloseOutputAsync(closeStatus, closeStatusDescription, CancellationToken);
```

CarnaViire commented on src/libraries/System.Net.WebSockets/tests/WebSocketCloseTests.cs:
```diff
@@ -0,0 +1,89 @@
+                WebSocketReceiveResult closing = await client.ReceiveAsync(new ArraySegment<byte>(receiveBuffer), CancellationToken);
+                Assert.Equal(WebSocketMessageType.Close, closing.MessageType);
+                Assert.Equal(closeStatus, closing.CloseStatus);
+                Assert.Equal(serverMessage, closing.CloseStatusDescription);
```
```suggestion
                Assert.Equal(closeStatusDescription, closing.CloseStatusDescription);
```

CarnaViire commented:
Could you please update your branch to the current main @IDisposable ? I saw build failures that seem to be caused by branch being out of date

IDisposable committed:
Use better local variable name

Renamed local `serverMessage` back to `closeStatusDescription` per PR feedback.
Co-authored-by: Natalia Kondratyeva <knatalia@microsoft.com>

IDisposable committed:
Address PR and rebase to main

Rebased to current main, updated the commit messages and added the remaining changes to address the PR comments.

IDisposable commented:
I've addressed all the PR comments and rebased to current main, but it won't build on my machine anymore... doing a Full build now to see if things will work, but figured I could push up and let the CI try,

I get this when building the .SLN in the directory (worked before today):

```
Error	MSB4018	The "Microsoft.DotNet.ApiCompat.Task.ValidateAssembliesTask" task failed unexpectedly.
System.MissingMethodException: Method not found: 'Void Microsoft.CodeAnalysis.CSharp.CSharpCompilationOptions..ctor(Microsoft.CodeAnalysis.OutputKind, Boolean, System.String, System.String, System.String, System.Collections.Generic.IEnumerable`1<System.String>, Microsoft.CodeAnalysis.OptimizationLevel, Boolean, Boolean, System.String, System.String, System.Collections.Immutable.ImmutableArray`1<Byte>, System.Nullable`1<Boolean>, Microsoft.CodeAnalysis.Platform, Microsoft.CodeAnalysis.ReportDiagnostic, Int32, System.Collections.Generic.IEnumerable`1<System.Collections.Generic.KeyValuePair`2<System.String,Microsoft.CodeAnalysis.ReportDiagnostic>>, Boolean, Boolean, Microsoft.CodeAnalysis.XmlReferenceResolver, Microsoft.CodeAnalysis.SourceReferenceResolver, Microsoft.CodeAnalysis.MetadataReferenceResolver, Microsoft.CodeAnalysis.AssemblyIdentityComparer, Microsoft.CodeAnalysis.StrongNameProvider, Boolean, Microsoft.CodeAnalysis.MetadataImportOptions, Microsoft.CodeAnalysis.NullableContextOptions)'.
   at Microsoft.DotNet.ApiSymbolExtensions.AssemblySymbolLoader..ctor(Boolean resolveAssemblyReferences)
   at Microsoft.DotNet.ApiSymbolExtensions.AssemblySymbolLoaderFactory.Create(Boolean shouldResolveReferences) in /_/src/Microsoft.DotNet.ApiSymbolExtensions/AssemblySymbolLoaderFactory.cs:line 12
   at Microsoft.DotNet.ApiCompatibility.Runner.ApiCompatRunner.CreateAssemblySymbols(IReadOnlyList`1 metadataInformation, ApiCompatRunnerOptions options, Boolean& resolvedExternallyProvidedAssemblyReferences) in /_/src/ApiCompat/Microsoft.DotNet.ApiCompatibility/Runner/ApiCompatRunner.cs:line 123
   at Microsoft.DotNet.ApiCompatibility.Runner.ApiCompatRunner.ExecuteWorkItems() in /_/src/ApiCompat/Microsoft.DotNet.ApiCompatibility/Runner/ApiCompatRunner.cs:line 49
   at Microsoft.DotNet.ApiCompat.ValidateAssemblies.Run(Func`2 logFactory, Boolean generateSuppressionFile, String[] suppressionFiles, String suppressionOutputFile, String noWarn, Boolean respectInternals, Boolean enableRuleAttributesMustMatch, String[] excludeAttributesFiles, Boolean enableRuleCannotChangeParameterName, String[] leftAssemblies, String[] rightAssemblies, Boolean enableStrictMode, String[][] leftAssembliesReferences, String[][] rightAssembliesReferences, Boolean createWorkItemPerAssembly, ValueTuple`2[] leftAssembliesTransformationPatterns, ValueTuple`2[] rightAssembliesTransformationPatterns) in /_/src/ApiCompat/Microsoft.DotNet.ApiCompat.Shared/ValidateAssemblies.cs:line 90
   at Microsoft.DotNet.ApiCompat.Task.ValidateAssembliesTask.ExecuteCore() in /_/src/ApiCompat/Microsoft.DotNet.ApiCompat.Task/ValidateAssembliesTask.cs:line 127
   at Microsoft.NET.Build.Tasks.TaskBase.Execute() in /_/src/Tasks/Common/TaskBase.cs:line 52
   at Microsoft.DotNet.ApiCompat.Task.ValidateAssembliesTask.Execute() in /_/src/ApiCompat/Microsoft.DotNet.ApiCompat.Task/ValidateAssembliesTask.cs:line 117
   at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute()
   at Microsoft.Build.BackEnd.TaskBuilder.<ExecuteInstantiatedTask>d__26.MoveNext()	System.Net.WebSockets (src\System.Net.WebSockets)	C:\Users\idisp\.nuget\packages\microsoft.dotnet.apicompat.task\8.0.100-preview.2.23107.1\build\Microsoft.DotNet.ApiCompat.ValidateAssemblies.Common.targets	16	
```

CarnaViire commented:
> doing a Full build now to see if things will work

Yeah, when you rebase, it's better to rebuild everything from scratch (`git clean -xdf` and then `.\build.cmd clr+libs -rc release`), otherwise the old binaries might not be compatible.

CarnaViire reviewed (approved):
LGTM, thanks!

CarnaViire commented:
I see that CLA check is stuck (it is a known issue).
I'm going to close and then reopen the PR to see if it helps.

CarnaViire commented:
Test failure is unrelated #83655

CarnaViire commented:
The chance of a backport is extremely low, unless there's a significant business impact. Given that currently there's only one hit, I assume you would need to switch to .NET 8 @chenxinyanc

IDisposable commented:
If, by chance, we do need a backport, I would love to learn the process to do that as well.
